agent_build_refactored/tools/runner.py

```diff
@@ -0,0 +1,910 @@
+            for file_path in self._tracked_files:
+                dest_path = isolated_source_root / file_path.parent.relative_to(
+                    SOURCE_ROOT
+                )
+                dest_path.mkdir(parents=True, exist_ok=True)
+                shutil.copy2(file_path, dest_path)
+
+            all_env_variables = self._get_all_environment_variables(work_dir=work_dir)
+            env_variables_str = "\n    ".join(
+                f"{n}='{v}'" for n, v in all_env_variables.items()
+            )
+            logging.info(
+                f"Start step: {self.id}\n"
+                f"Passed env. variables:\n    {env_variables_str}\n"
+            )
+            try:
+                self._run_script(work_dir=work_dir)
+            except Exception:
+                globs = [str(g) for g in self.tracked_files_globs]
+                logging.exception(
+                    f"'{type(self).__name__}' has failed. "
+                    "HINT: Make sure that you have specified all files. "
+                    f"For now, tracked files are: {globs}."
+                )
+                raise
+            finally:
+                self.cleanup()
+        else:
+            log.info(f"Result if the step '{self.id}' is found in cache, skip.")
+
+        self._save_to_cache(
+            is_skipped=skipped,
+            output_directory=output_directory,
+            cache_directory=cache_directory,
+        )
+
+    def cleanup(self):
+        if self._step_container_name:
+            check_call_with_log(["docker", "rm", "-f", self._step_container_name])
+
+    def __eq__(self, other: "RunnerStep"):
+        return self.id == other.id
+
+
+class ArtifactRunnerStep(RunnerStep):
+    """
+    Specialised step which produces some artifact as a result of its execution.
+    """
+
+    def _restore_cache(
+        self, output_directory: pl.Path, cache_directory: pl.Path
+    ) -> bool:
+        self._remove_output_directory(output_directory=output_directory)
+
+        if cache_directory.exists():
+
+            output_directory.symlink_to(cache_directory)
+            return True
+
+        return False
+
+    def _save_to_cache(
+        self, is_skipped: bool, output_directory: pl.Path, cache_directory: pl.Path
+    ):
+        if not is_skipped:
+            shutil.copytree(output_directory, cache_directory, dirs_exist_ok=True)
+
+
+class EnvironmentRunnerStep(RunnerStep):
+    """
+    Specialised step which performs some actions on some environment in order to prepare if for further uses.
+        If this step runs in docker, it performs its actions inside specified base docker image and produces
+        new image with the result environment.
+        If step does not run in docker, then its actions are performed directly on current system.
+    """
+
+    def _restore_cache(
+        self, output_directory: pl.Path, cache_directory: pl.Path
+    ) -> bool:
+        if self.runs_in_docker:
```

Comment:
Perhaps to make it a bit easier to read by reducing one level of indentation you could inverse the condition and return ``False`` in case ``self.runs_in_docker`` is false.